Add ":curwin" command modifier that makes split commands use current window

ychin opened the pull request:
Add a new command modifier that changes how splits work, similar to `:vertical` and `:tab`. The difference is that it will make the command use an *existing* window rather than splitting off a new window.

The user can use `[count]` to specify which window to use (defaults to current window) and [!] to forcefully discard a window with modified buffer.

Most of the handling happens in `win_split()` where it detects the command mod and pretends to split off a window. Commands like `:terminal` handles the modifier manually because it already has to deal with the `++curwin` option.

Examples of commands affected by this:

* help
* split / new
* sbuffer
* snext
* ptag
* cw / copen
* cc / cnext
* terminal

# Rationale

Just for some rationale and background for why I implemented this.

1. I find that a lot of times I do something similar to `:help | only` or `:tab help` just to get a help page taking up a full screen. It seems to me there should just be a command that says "use the current window I have and show the help page there", instead of always having to split a window. Of course you can have two commands: `:shelp` and `:help` but that's redundant, and a better solution is to have a modifier that works across the board similar to how `:vertical` works.

2. Similarly I have a custom command called `:EditConfig` that opens my vimrc in a split window. Again, sometimes I want to *just* open the vimrc without the split and I don't want to have to write two commands, or have to do `:only` afterwards.

3. I have some Vim scripts that augment `mksession` and save out quickfix content (by calling `getqflist()` and `getloclist()` on all windows) into the extra `x.vim`. However, I found it really hard for Vim scripts to properly restore the quickfix windows because the session restore will have created the window splits for me, but there's no way for me to just say "window 5 is the location list display for window 3" without some complicated command scripting. Now, I can just add this line to do that: `3wincmd w | 5curwin lopen`.

* Thought: I do think it would be nice for Vim to have a "quickfix" option in `sessionoptions` that will do that natively.


# Existing bug in Vim

Note that there's currently a bug / issue where the `<mods>` passed into custom commands don't retain `[count]` and `[!]` for modifiers, so you cannot do something like `:2curwin! MyCommand` as `<mods>` will just show "curwin". This is probably something that should be fixed. It also affects `:tab`. I think the bug is in `uc_check_code()`.

# Tests
I have not implemented any tests yet, but I could. I couldn't find any tests for existing command modifiers like `:vert` and `:tab` so I didn't have a good place to add them, but I would imagine for new functionality like this that touches a fair bit of code I should probably add some basic unit tests.

Edit: Unit tests are now added in 59671a217208c9b417f3293c62ea01a71247cc54.

# Other names

I considered naming this `usewin` instead of `curwin`. Still not sure which one is better.

# Ideas for other similar commands

Following down this path, just some ideas for similar modifiers:

* `:[count]usetab`: Go to a particular tab specified by `[count]` and make the split there. Unlike `:tab` this doesn't make a new tab. Can combine with `:vertical` and `:curwin`, but not `:tab`.

* `:[count]abovewin`: Make a split above window specified by `[count]`, kind of like how `:tab` makes a new tab next to a specified one. This is more flexible than `:topleft` (same as `:0abovewin`) and `:botright` (same as `:$abovewin`)

ychin committed:
Add ":curwin" modifier to Vim

Add a new command modifier that changes how splits work, similar to
`:vertical` and `:tab`. The difference is that it will make the command
use an *existing* window rather than splitting off a new window.

The user can use [count] to specify which window to use (defaults to
current window) and [!] to forcefully discard a window with modified
buffer.

Most of the handling happens in `win_split()` where it detects the
command mod and pretends to split off a window. Commands like
`:terminal` handles the modifier manually because it already has to deal
with the `++curwin` option.

Examples of commands affected by this:

* help
* split / new
* sbuffer
* snext
* ptag
* cw / copen
* cc / cnext
* terminal

Note that there's currently a bug / issue where the "<mods>" passed into
custom commands don't retain [count] and [!] for modifiers, so you
cannot do something like `:2curwin! MyCommand` as "<mods>" will just
show "curwin". This is something that should be fixed. It also affects
`:tab`.

ychin commented:
Seems like I broke unit tests. Somehow missed that. Let me fix that first

ychin commented:
Ok I believe I have fixed up everything from unit tests to documentations. Should be good to go now.

k-takata commented on runtime/doc/map.txt:
```diff
@@ -1441,11 +1441,11 @@ The valid escape sequences are
-		|:keepjumps|, |:keepmarks|, |:keeppatterns|, |:leftabove|,
-		|:lockmarks|, |:noswapfile| |:rightbelow|, |:silent|, |:tab|,
-		|:topleft|, |:verbose|, and |:vertical|.
+                nothing. Supported modifiers are |:aboveleft|, |:belowright|,
```
Tab should be used in the help.

ychin replied:
Ok.  Just pushed a new commit with tabs.

Just wondering though. Should help docs have `noet` in the modeline at the end of file? It already sets `ts=8`.

brammool commented:
Yee Cheng Chin wrote:

> Add a new command modifier that changes how splits work, similar to
> `:vertical` and `:tab`. The difference is that it will make the
> command use an *existing* window rather than splitting off a new
> window.
> 
> The user can use `[count]` to specify which window to use (defaults to
> current window) and [!] to forcefully discard a window with modified
> buffer.
> 
> Most of the handling happens in `win_split()` where it detects the
> command mod and pretends to split off a window. Commands like
> `:terminal` handles the modifier manually because it already has to
> deal with the `++curwin` option.
> 
> Examples of commands affected by this:
> 
> * help
> * split / new
> * sbuffer
> * snext
> * ptag
> * cw / copen
> * cc / cnext
> * terminal
> 
> # Rationale
> 
> Just for some rationale and background for why I implemented this.
> 
> 1. I find that a lot of times I do something similar to `:help | only`
> or `:tab help` just to get a help page taking up a full screen. It
> seems to me there should just be a command that says "use the current
> window I have and show the help page there", instead of always having
> to split a window. Of course you can have two commands: `:shelp` and
> `:help` but that's redundant, and a better solution is to have a
> modifier that works across the board similar to how `:vertical` works.
> 
> 2. Similarly I have a custom command called `:EditConfig` that opens
> my vimrc in a split window. Again, sometimes I want to *just* open the
> vimrc without the split and I don't want to have to write two
> commands, or have to do `:only` afterwards.
> 
> 3. I have some Vim scripts that augment `mksession` and save out
> quickfix content (by calling `getqflist()` and `getloclist()` on all
> windows) into the extra `x.vim`. However, I found it really hard for
> Vim scripts to properly restore the quickfix windows because the
> session restore will have created the window splits for me, but
> there's no way for me to just say "window 5 is the location list
> display for window 3" without some complicated command scripting. Now,
> I can just add this line to do that: `3wincmd w | 5curwin lopen`.
> 
> * Thought: I do think it would be nice for Vim to have a "quickfix"
> option in `sessionoptions` that will do that natively.

Hmm, for most of the commands there are non-splitting variants.  It's
like for those there is a "split" modifier, but it's part of the command
name.  Thus the "curwin" modifier is only useful for commands that split
by default.  And :help would be the most prominent one.  :terminal is
another one, for which we have the ++curwin argument.

So why not use :help ++curwin ?


-- 
        **  Hello and Welcome to the Psychiatric Hotline **
If you are obsessive-compulsive, please press 1 repeatedly.
If you are co-dependent, please ask someone to press 2.
If you have multiple personalities, please press 3, 4, 5 and 6.
If you are paranoid-delusional, we know who you are and what you want
       - just stay on the line so we can trace the call.
If you are schizophrenic, listen carefully and a little voice will
       tell you which number to press next.
If you are manic-depressive, it doesn't matter which number you press
       - no one will answer.
If you suffer from panic attacks, push every button you can find.
If you are sane, please hold on - we have the rest of humanity on the
        other line and they desparately want to ask you a few questions.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

ychin commented:
> It's like for those there is a "split" modifier, but it's part of the command name. 

I think with how most command modifiers work (e.g. `vertical` and `tab`), currently `split` is the de facto "open a new window" command that lets other modifiers handle how to deal with it. Otherwise there is no general command that works this way, e.g. `:e some_file` is not modifiable in anyway.

Here are some extra points I can think of why having a dedicated `:curwin` modifier is desirable:

1. `[count]` adds unique capability that cannot be done with a simple `++curwin` option, since it lets you choose the placement of the window. This is especially important for commands that makes a new window that are *context sensitive* to the current window, and it's hard to make those commands show the window in the right place. E.g. let's say I'm in the following:

   ```
   +---+---+---+
   |[1]| 2 | 3 |
   +---+---+---+
   |     4     |
   +-----------+
   ```
   
   I'm editing `foo.c` in window 1, and I want to:
     - open the header `foo.h` in window 2,
     - also open current file (`foo.c`) in 3, and
     - open the location list in 4.
   
   Right now I have to do a lot of complex `wincmd` dances to make it work. But now I can just do the following: (Note that `lopen`, `%`, and `sp` are all context sensitive to current window so you can't just move to window 2/3/4 first and do the commands)
   
   ```
   :+curwin sp %:s/.c/.h/
   :wincmd p
   :3curwin sp
   :wincmd p
   :$curwin lopen
   ```
   
   As I mentioned in the "Rationale" part in original comment, this is also very useful in scripts as it lets you precisely place your windows especially when combined with `mksession`.

2. `[!]` also adds a global way of incidating you want to discard a file. `++curwin` option currently doesn't work that way. You can already do `:terminal! ++curwin` but this is actually undocumented, and `:help!` already means something else. Being able to say `:curwin!` fixes this problem and is consistent with how Vim commands generally work.

3. It works with `<mods>`. Plugin authors don't need to think about exposing duplicate functionality for every command they make. It "just works" if they already support other modifiers. Only caveat is as I mentioned in original post that `[count]` and `!` are not supported in `<mods>` right now, which is a problem already for `:[count]tab`.

4. It will be standardized and not needing individual options added to each command. This is similar to `vert` and `tab` where you don't have to remember how to invoke individual options with a different syntax.

This does create some redundancy in commands (`:curwin sp foo.txt` vs `:e foo.txt`, `:curwin term` vs `:term ++curwin`), but I think it's ok. It allows you to approach a command in different ways, and we already have `:vsp` vs `:vert sp` or `:tab new` vs `:tabnew`.

> So why not use :help ++curwin ?

In addition to `terminal` and `help`, I think the following commands would also be helped by this (this re-iterates the above points):

* `split` (no arguments)
* `lopen` / `copen`
* User commands with `<mods>`
* `expand()` and `cmdline-special` (since they are context sensitive to current window and you can use `[count]curwin` to apply the current command to another window)

ychin commented:
Just rebased the code.

Also some thoughts on the how the local window state would be preserved. Right now when you split a new window (e.g. `:split`) the new window inherits the old one's location list, local dir, etc. `[count]curwin` doesn't do that currently. So when you do `+curwin sp` it will populate the next window with the current window's content, but it won't make the next window have the exact same local properties.

I think the current behavior is ok, but we could also make `[count]curwin` resets the target window's local properties to the current one too if we really want to.

brammool commented:
I see no positive responses or thumbs-up on this PR and I don't think it has a clear usefulness.  Therefore closing.

ychin commented:
I have actually been using this custom command almost daily combined with `mksession` and found it incredibly useful in being able to restore various states like quickfix and location lists in the correct windows, along the ability to do `:curwin help` to display help page in the current window. Oh well. I hope some other users will find this PR and comment on it.

justinmk commented:
I would think this makes more sense than adding `++curwin` to every command.

Seems like Vim cannot decide if commands should be prefixed or suffixed.
